samples/Mvc.DynamicCorsPolicyProvider/DynamicCorsPolicyProvider.cs

```diff
@@ -0,0 +1,94 @@
+using Microsoft.AspNetCore.Cors.Infrastructure;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using Newtonsoft.Json;
+
+namespace Mvc.DynamicCorsPolicyProvider
+{
+    public class DynamicCorsPolicyProvider : ICorsPolicyProvider
+    {
+        private readonly PhysicalFileProvider _fileProvider;
+        private readonly ILogger<DynamicCorsPolicyProvider> _logger;
+        private Dictionary<string, CorsPolicy> _policies = null;
+
+        public DynamicCorsPolicyProvider(PhysicalFileProvider fileProvider, ILogger<DynamicCorsPolicyProvider> logger)
+        {
+            _fileProvider = fileProvider;
+            _logger = logger;
+
+            // Initialize the policies the first time this policy provider is instantiated (singleton)
+            BuildPolicies();
+
+            // Register a file change callback to refresh these policies whenever a change is made
+            RegisterCallback();
+        }
+
+        public Task<CorsPolicy> GetPolicyAsync(HttpContext context, string policyName)
+        {
+            return Task.FromResult(_policies[policyName]);
+        }
+
+        private void RegisterCallback()
+        {
+            _fileProvider.Watch("corsSettings.json")
+                .RegisterChangeCallback((state) =>
+                {
+                    _logger.LogInformation("CORS settings changed. Updating policies...");
+                    BuildPolicies();
+
+                    // Re-register callback
+                    RegisterCallback();
+                }, state: null);
+        }
+
+        private void BuildPolicies()
+        {
+            var content = File.ReadAllText(_fileProvider.GetFileInfo("corsSettings.json").PhysicalPath);
+            var policySettings = JsonConvert.DeserializeObject<IEnumerable<PolicySetting>>(content);
+
+            var temp = new Dictionary<string, CorsPolicy>(StringComparer.OrdinalIgnoreCase);
+            foreach (var policySetting in policySettings)
+            {
+                var corsPolicyBuilder = new CorsPolicyBuilder();
+                if (policySetting.AllowedOrigins == null || policySetting.AllowedOrigins.Length == 0)
+                {
+                    corsPolicyBuilder.AllowAnyOrigin();
+                }
+                else
+                {
+                    corsPolicyBuilder.WithOrigins(policySetting.AllowedOrigins);
+                }
+
+                if (policySetting.AllowedHeaders == null || policySetting.AllowedHeaders.Length == 0)
+                {
+                    corsPolicyBuilder.AllowAnyHeader();
+                }
+                else
+                {
+                    corsPolicyBuilder.WithHeaders(policySetting.AllowedHeaders);
+                }
+
+                temp.Add(policySetting.Name, corsPolicyBuilder.Build());
+            }
+
+            _logger.LogInformation("Updating existing CORS policies...");
```

Comment:
Given that there's already a log message above saying that something changed and it's reloading, I suggest moving this log message to the very end of this method and say `"Updat**ed** CORS policies"` to signify that the operation has completed.